### DISCUSSION TITLE
style.load alternative?

### DISCUSSION THREAD

**Discussion by housec84** [maplibre/maplibre-gl-js#2716] [General]
Hi, I'm currently migrating from mapbox to maplibre - great work all whom are involved!

However come across an issue, where changing styles blows up our application. I think the culprit is the style.load event isn't being fired for style loads anymore...?

`            mapLocal.on('style.load', function () {
                console.log("style.load"); 
                setStyleChange(a => !a);
            });`

This was based off the documentation from mapbox here:  https://docs.mapbox.com/mapbox-gl-js/example/style-switch/
(this page isn't in maplibre docs).

the Style.load event also isn't present in the mallibre api docs:
https://docs.mapbox.com/mapbox-gl-js/api/map/#map.event:style.load
https://maplibre.org/maplibre-gl-js-docs/api/map/#map.event:style.load

Is there an alternative method of re-adding data driven sources/layers on style changes that is now recommended?

Apologies if this has been covered before - a through search on my part hasn't yielded anything.

**Comment by acalcutt** [comment]
I also had this issue with the way I was changing styles and re-loading layers on 'style.load'. This has been discussed a bit in https://github.com/maplibre/maplibre-gl-js/issues/2587

One thing I got to work was using [SetStyle](https://maplibre.org/maplibre-gl-js-docs/api/map/#map#setstyle) transformStyle option to get away from using 'style.load'. 

I was able to change the custom sources and layers to 'load' instead of 'style.load' (which only runs once), then with the SetStyle transformStyle options I was able to move those sources and layers over to the new style when SetStyle was run. I ended up with a SetStyle code something like this. ([example page](https://wifidb.net/demo/trails/trails.html))

```
map.setStyle('https://tiles.wifidb.net/styles/' + styleId + '/style.json', {
	transformStyle: (previousStyle, nextStyle) => ({
		...nextStyle,
		sources: {
			...nextStyle.sources,
			track_line: previousStyle.sources.track_line,
			contour_source: previousStyle.sources.contour_source,
			atv_trails: previousStyle.sources.atv_trails,
			atv_points: previousStyle.sources.atv_points
		},
		layers: [
			...nextStyle.layers.filter(layer =>  layer.type !== 'symbol'),
			...previousStyle.layers.filter(layer =>  layer.id === 'contours'),
			...previousStyle.layers.filter(layer =>  layer.id === 'contour_text'),
			...previousStyle.layers.filter(layer =>  layer.id === 'atv_trails'),
			...previousStyle.layers.filter(layer =>  layer.id === 'track_line_layer'),
			...nextStyle.layers.filter(layer =>  layer.type === 'symbol'),
			...previousStyle.layers.filter(layer =>  layer.id === 'atv_points'),
		]
	})
});
```

This wasn't perfect since it reordered the layers a bit, but with the filtering I was at least able to order the layers a little bit (moving symbol layers mostly to the top). 

It did seem to make the style switching a bit faster, because it no longer seems like it needs to recreate those layers, it just moves them to the new style.

**Reply by housec84** [reply]
Ah brilliant, that looks like it would work.  I'll give that a go. 
Looks better than the hack I was working on putting in with a setTimeout over map.isStyleLoaded call.
Thanks!!

**Reply by housec84** [reply]
Can confirm. works like a charm... 
And as you say, it does appear to be faster too.
Many thanks!
